aya/src/programs/mod.rs

```diff
@@ -386,6 +386,18 @@ impl<T: Link> ProgramData<T> {
     }
 }
 
+fn unload_program<T: Link>(data: &mut ProgramData<T>, detach: bool) -> Result<(), ProgramError> {
```

Comment:
I'm not sure that `detach` should not be an option here.

There are a few reasons for this:
1. I don't think `unload(data, false)` will actually unload the program if it's attached... as any links created will still reference the program fd... so the kernel won't remove it. Will have to test it out to be sure.
1. If you `unload(data, false)` and `load()` again, any links that were not detached are still pointing at the old program fd which is a bit weird

Reply:
`unload(false)` means "unload the program once the links are detached" (presumably manually). Or in other words, it's "decrement_program_refcount(program)". 

Even without the boolean arg, it's still not guaranteed that `unload()` will unload a program, since you might have owned links.  I think `unload(bool)` is fine.

Reply:
That still leaves `2.` in my list, which is what happens if you call `unload(false)`, followed by `load()`

Because `Program` owns `Links` and the fd has changed (thus the links are not really for this program, but the old version), what do we do here?

Keep them? Feels weird. Especially as `attach()` might fail because the old link will keep the old program alive... and that link may (in the case of XDP for example) prevent attach.

Drop them all? In which case, we could just make that default and there's no need for the bool

Something else?...

Reply:
> That still leaves `2.` in my list, which is what happens if you call `unload(false)`, followed by `load()`
> 
> Because `Program` owns `Links` and the fd has changed (thus the links are not really for this program, but the old version), what do we do here?

Agreed this is weird.

> Keep them? Feels weird. Especially as `attach()` might fail because the old link will keep the old program alive... and that link may (in the case of XDP for example) prevent attach.

This is just how attach works: it can fail and in many cases you have to handle possible failures unless you completely control the whole system. Take the xdp case - another process might have attached something, or someone via iproute, or the driver might not support your attach mode, etc. I don't think this is a problem.

>  Drop them all? In which case, we could just make that default and there's no need for the bool

Yes now I think we should do this, and document that if you have owned links, those will stay attached and will keep the old instance of the program loaded.